### ISSUE TITLE
Controller delete volume is called before node unstage volume returns success

### ISSUE THREAD

**Issue by rvakash** [kubernetes/kubernetes#94108]
**What happened**:
We are seeing an issue where Controller issued Controller Delete volume while node unstage volume was still in flight but unpublish volume had returned success. This is a block volume on K8s 1.18.2. Is this the desired behavior ? Or is there a bug open on this ? According to the spec, my understanding was controller delete volume would be issued after node unstage volume returned success
The problem is unstage volume makes a call to the backend storage provider but since delete volume was issued before that. The volume gets deleted in the backend and the unstage volume (or umount) never finishes. We are left with stale devices on the K8s nodes

**What you expected to happen**:
Controller delete volume to be issued only after node unstage volume returns success

**How to reproduce it (as minimally and precisely as possible)**:
I think its a race condition. Not sure how to reproduce it.

**Anything else we need to know?**:
NodeUnpublish volume does - $umount on TargetPath and then $rm -rf on TargetPath. Possible that K8s is looking at this mount point and decided unstage happened and issued delete volume. 

**Environment**:
- Kubernetes version (use `kubectl version`):1.18.2

**Log snippets**:
- Unpublish success:  
Aug 18 23:08:41 c3-k8s-n5 kubelet[998]: I0818 23:08:41.303322 998 operation_generator.go:1145] UnmapVolume succeeded for volume "kubernetes.io/csi/csi-hxcsi^2b6fdb11-ebbd-4847-a089-e3d846cc808a" (OuterVolumeSpecName: "test-volume-block") pod "e4bbf342-a446-4fb3-bd7f-ab5c17fe269f"

- Unstage invoked:
Aug 18 23:08:41 c3-k8s-n5 kubelet[998]: I0818 23:08:41.389774 998 reconciler.go:312] operationExecutor.UnmountDevice started for volume "pvc-98407c6c-9e3c-4d27-8728-1859f9cffde8" (UniqueName: "kubernetes.io/csi/csi-hxcsi^2b6fdb11-ebbd-4847-a089-e3d846cc808a") on node "c3-k8s-n5"

- Unstage failed: 
Aug 18 23:08:43 c3-k8s-n5 kubelet[998]: E0818 23:08:43.238461 998 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/csi-hxcsi^2b6fdb11-ebbd-4847-a089-e3d846cc808a podName: nodeName:}" failed.

- Delete Volume invoked:
{"log":"I0819 06:08:41.821406 1 controller.go:1311] delete \"pvc-98407c6c-9e3c-4d27-8728-1859f9cffde8\": started\n","stream":"stderr","time":"2020-08-19T06:08:41.821566667Z"}

**Comment by msau42** [comment]
I think the kubelet logic that determines when to delete the Pod object (which storage protection uses as a signal that unmount is done) doesn't handle staging paths.

**Comment by msau42** [comment]
@kubernetes/sig-storage-bugs

**Comment by travisghansen** [comment]
In my case, unstage isn't even invoked first...it's just bad order of operations.

**Comment by travisghansen** [comment]
Here's my observations so far...scenario, I'm creating 2 pvcs and 2 pods attaching. I then delete the pods, wait for that to finish, then I delete the PVCs.

In 1.17.4 when the pods are deleted, Unpublish is invoked and subsequently Unstage (without deleting pvc)
In 1.18.6 when the pods are deleted, Unpublish is invoked, Unstage is not invoked until I delete the PVCs. But when I delete the PVCs first controller Delete is invoked, then Unstage on the node.

**Comment by msau42** [comment]
That is really odd. Unstage should only depend on Pod deletion not PVC deletion.

**Comment by travisghansen** [comment]
@msau42 yeah, I just deleted the PVCs in the 1.18 cluster after letting it sit for 1.5 hours post pod deletion, and immediately experienced exactly what I described above. 0 unstage calls were invoked during the 1.5 hour window :( Now they fail due to stale file handles (nfs-based volumes) because delete is invoked before unstage and it's impossible to make the unmount go through.

**Comment by msau42** [comment]
Can you check your kubelet.log to see if there's any errors that could indicate why Unstage wasn't called?

**Comment by travisghansen** [comment]
@msau42 I don't see anything. If there was an issue preventing it from calling unstage would that also prevent it from invoking unstage after the PVC is deleted as well?

**Comment by saikat-royc** [comment]
@travisghansen  Can you also share the details what is the version of the external provisioner sidecar that is used? Also is this a dynamically provisioned volume?

**Comment by travisghansen** [comment]
@saikat-royc in both cases it's using `image: quay.io/k8scsi/csi-provisioner:v1.6.0`

**Comment by jingxu97** [comment]
@travisghansen a few questions
1. which csi driver are you usinig?
2. kubelet side UnmountDevice  is issued, so normally csi driver should trigger NodeUnstageVolume call?
3. seems like your pod cannot be deleted for a long time. When you try to delete PVC, what is the status of the pod? Also do PVC and PV objects get deleted?

**Comment by travisghansen** [comment]
1. My own implementation here https://github.com/democratic-csi/democratic-csi
2. I'm not sure what the question is here. From the csi side unpublish is invoked (and succeeds)...unstage does not get invoked on 1.18.6 until pvc is deleted and, sadly, **after** the controller delete.
3. My last test the pods were successfully deleted for an 1.5 hours (and unpublish succeeded immediately). So the pods were deleted for a substantial amount of time (and unstage *not* invoked for a substantial amount of time) until I deleted the PVC. When I delete the PVC (via kubectl) the PVC gets deleted immediately. The PV gets deleted I'm guessing because the controller delete call is invoked and succeeds. Kubelet appears to notice that the volume is still staged (and possibly combined with the fact that the PV is now gone) and starts to invoke unstage and continues to do so since it's failing.

Those are my observations to this point.

I'm currently testing with a very basic sub-driver of my project `nfs-client`. I'll also re-iterate, the same driver (and various 'external-foo') components/versions are installed in both clusters. The difference appears to be the cluster version as 1.17 is behaving as expected with the exact same set of components.

**Comment by jingxu97** [comment]
@travisghansen  For question 2, because I see the log you provided 
```
Unstage invoked:
Aug 18 23:08:41 c3-k8s-n5 kubelet[998]: I0818 23:08:41.389774 998 reconciler.go:312] operationExecutor.UnmountDevice started for volume "pvc-98407c6c-9e3c-4d27-8728-1859f9cffde8" (UniqueName: "kubernetes.io/csi/csi-hxcsi^2b6fdb11-ebbd-4847-a089-e3d846cc808a") on node "c3-k8s-n5"

Unstage failed:
Aug 18 23:08:43 c3-k8s-n5 kubelet[998]: E0818 23:08:43.238461 998 nestedpendingoperations.go:301] Operation for "{volumeName:kubernetes.io/csi/csi-hxcsi^2b6fdb11-ebbd-4847-a089-e3d846cc808a podName: nodeName:}" failed.
```

This means that kubelet invokes UnmountDevice right after UnmapVolume, which is the normal workflow. UnmountDevice should trigger csi.NodeUnstageVolume on the driver. So I don't why the driver does not receive this request. Could you check your driver log that during this period of time, what logs did it have?

**Comment by travisghansen** [comment]
Wires are getting crossed. I didn’t submit this issue so those are not my logs. My problem is similar but slightly different.

In my case unstage is not getting called at all until pvc delete which is slightly different from the author who is saying that unstage is getting invoked but delete is not waiting for it to complete successfully.

Sorry for the confusion. We can create a new issue if that’s a better approach.

**Comment by jingxu97** [comment]
@travisghansen i see. then pls go ahead to open another issue and give us details about your issue. Thanks!

**Comment by gnufied** [comment]
I think what @travisghansen is seeing might be different thing but what @ravkash reported originally sounds like that is how it will work. I wonder if we should check for `volumeInUse` fields in pvc_protection controller, so as we don't delete PVCs until they are not in-use.

**Comment by jingxu97** [comment]
@saikat-royc and I are discussing the possible fix using volumeInUse fields now. Need to change certain things to make it work.

**Comment by saikat-royc** [comment]
I am actively working on a fix for the race between node unstage and volume delete.

**Comment by Jiawei0227** [comment]
Trying to understand more on this, is it possible to reject a DeleteVolume call if the volume is still staged? Are we able to somehow do that?

**Comment by msau42** [comment]
@saikat-royc is working on a solution in external-provisioner to check for nodeunstage success.

**Comment by Jiawei0227** [comment]
> @saikat-royc is working on a solution in external-provisioner to check for nodeunstage success.

Yeah, I am thinking more about from the CSI perspective. When driver deletes volume, is it able to check whether the volume is being staged on some node?

**Comment by msau42** [comment]
Not all drivers will have that ability, especially if the process of mounting something on a node doesn't show in the control plane.

**Comment by aksatish19** [comment]
Hello, was this issue ever fixed?
